### PR TITLE
✨ Update score for branch protection with levels

### PR DESCRIPTION
**Pull request by laurentsimon**
Update score to work like 'stages':
- first stage: force push, force deletion. if admin, enforce admins. Those give 3/10
- second stage: review >= 1. for admins, up-to-date branch. Those gives 6/10. Maybe up-to-date branch could go to first stage?
- third stage: status checks defined. Gives 8/10. 
- fourth stage : reviews >= 2. Gives 9/10
- fifth stage: for admins, dismissal (very hard to use). Gives 10/10

If a stage is not satisfied, the score won't go up regardless of the settings of the next stages. Warn/Info messages are shown for all stage, regardless of the score.
I scorecard is run without admin access, the admin settings are ignored.

@olivekl how shall I word this in the checks.yaml?

FYI, I'm working on the unit tests right now but that should not hamper the review.

### CODE REVIEW

**Comment by laurentsimon** [comment]
friendly ping. ISE would like this check in the cron job's data in a couple weeks.

File: checks/branch_protection.go
Comment (asraa) [inline checks/branch_protection.go:33]: maybe add a comment
`// Points incremented at each level`

File: checks/branch_protection.go
Comment (azeemshaikh38) [inline checks/branch_protection.go:446]: Move this fn to `_test.go` file where it is used.

File: checks/branch_protection.go
Comment (asraa) [inline checks/branch_protection.go:132]: I think I'm misunderstanding this logic :/ why is it the case that across all branches the settings are the same or one is 0? 

Is it because the scores are binary (at least for now, given the values in `branchProtectionSettingScores`)?

**Review by asraa — COMMENTED** [review]
going to also comment offline as well..

File: checks/branch_protection.go
Comment (laurentsimon) [reply checks/branch_protection.go]: good call! Done. Thanks!

File: checks/branch_protection.go
Comment (laurentsimon) [reply checks/branch_protection.go:33]: done, thanks!

File: checks/branch_protection.go
Comment (laurentsimon) [reply checks/branch_protection.go:132]: the name of the function may be confusing. This is not the result score: it's the maximum score possible for a level, calculated by adding `+1` for each settings enabled. The maximum score depends on whether you have admin access or not. When you don't have admin access, the max score for a level may be 0 because the settings are not available. I don''t want to say "we've removed points because we cannot access your admin-only settings".
So in this function, I update the maximum possible score but check that it's not 0, ie that is it available.
As yoou say, the maximum score across all branches MUST be the same: I added this code to also verify that the maximum possible score is the same across all branches: if this is violated it means there's a bug in our implementation.
Does this make sense?

Comment (laurentsimon) [reply checks/branch_protection.go:132]: I think you're right there is unecessary logic I forgot to remove. let me update

Comment (laurentsimon) [reply checks/branch_protection.go:132]: done, PTAL

**Comment by olivekl** [comment]
Suggestions for wording this in checks.yaml (use `|` before to preserve formatting):

This test has tiered scoring. Each tier must be fully satisfied to achieve points at the next tier. For example, if you fulfill the Tier 3 checks but do not fulfill all the Tier 2 checks, you will not receive any points for Tier 3. 

Note: If Scorecard is run without an administrative access token, the requirements that specify “For administrators” are ignored. 

Tier 1 Requirements (3/10 points):

- Force push
- Force deletion
- For administrators: Include administrators

Tier 2 Requirements (6/10 points):

-  Required reviewers >=1 ​
- For administrators: Strict status checks (require branches to be up-to-date before merging) 

Tier 3 Requirements (8/10 points):

- Status checks defined

Tier 4 Requirements (9/10 points):

- Required reviewers >= 2

Tier 5 Requirements (10/10 points):

- For administrators: Dismiss stale reviews

File: checks/branch_protection.go
Comment (asraa) [inline checks/branch_protection.go:64]: I don't think this one is gated, is it? the scorecard check will use the settings I think because of this comment ("Even though it technically is not enforced by GitHub if Context=nil, we still show a positive outcome for users.")

File: checks/branch_protection.go
Comment (asraa) [inline checks/branch_protection.go:60]: suggest revising this comment to "Used to check if required status checks before merging is non-empty, rather than checking if requireStatusChecks is set." or something like that

without the history that there used to be `requireStatusChecks` I find it confusing.

File: checks/branch_protection.go
Comment (asraa) [inline checks/branch_protection.go:80]: suggest comment
`// maximum score depending on whether admin token is used`

File: checks/branch_protection.go
Comment (asraa) [inline checks/branch_protection.go:39]: could you map the levels to the names you've given them? 
1: (admin/non-admin) basic score
2. (admin/non-admin) review score
3. (non-admin) context
4. (non-admin) thorough review
5. (admin) thorough review

File: checks/branch_protection.go
Comment (asraa) [reply checks/branch_protection.go]: Ah! Gotcha. Was just a verification /debug check. Got it with the removed logic. Thank you!

File: checks/branch_protection.go
Comment (azeemshaikh38) [inline checks/branch_protection.go:403]: How about we write it this way:

```
tier1Score, tier1Passing := calculateTier1Score(branch.BranchProtectionRule)

if tier1Passing {
  tier2Score, tier2Passing := calculateTier2Score(...)
}
if tier2Passing {
  ...
}
score := tier1Score + tier2Score + ...
```

wherein, the `calculateTierXScore` returns true either if:
(i) all requirements for the tier are met, or
(ii) some fields are `nil`, indicating no data, so we don't penalize the repo.

This way we avoid the complex structs and also avoid leaking `admin token info` here in the check. Wdyt?

**Review by azeemshaikh38 — CHANGES_REQUESTED** [review]

File: checks/branch_protection.go
Comment (laurentsimon) [reply checks/branch_protection.go:403]: We can do the tier calculation in this part, because it's iterating over all branches: we need the data from all branches to calculate the tiers. This function is collecting the result. The tier calculation is done the `computeScore()` function. Can you check it?

Comment (asraa) [reply checks/branch_protection.go:403]: > (ii) some fields are nil, indicating no data, so we don't penalize the repo.

I think without overhauling and separating out score/check (maybe not in this PR, i'm happy to give this a tackle on monday), you will need to propogate a max score since nil will either mean nil or no admin token. so i'm not sure this will work unless you get that info, in which case i think it's pretty much equivalent to what's currently implemented

File: checks/branch_protection.go
Comment (laurentsimon) [reply checks/branch_protection.go:39]: sure

File: checks/branch_protection.go
Comment (azeemshaikh38) [reply checks/branch_protection.go:403]: > We cannot do the tier calculation in this part, because it's iterating over all branches: we need the data from all branches to calculate the tiers. This function is collecting the result. The tier calculation is done in the computeScore() function and essentially does what you describe here. Can you check it?

Ah, I see. Then, let's calculate per-branch tier scores here and aggregate them later? So each branch will have:

```
{
  tier1: <score, passing>,
  tier2: <score, passing>,
 ...
}
```

which will be later aggregated across all branches. Would that work?

Comment (azeemshaikh38) [reply checks/branch_protection.go:403]: > nil will either mean nil or no admin token

Right. And there is no way for us differentiate between the 2. Even with the knowledge that some fields are `admin-only`, a `nil` does not mean that it is because a lack of admin privileges. It simply means there was no data. So I see no reason to make speculations about why certain fields were returned `nil`.

File: checks/branch_protection.go
Comment (laurentsimon) [reply checks/branch_protection.go:64]: right. When I tested, I think this setting was always nil when Context=nil. So I think my comment no longer applies. Let me remove it.

File: checks/branch_protection.go
Comment (laurentsimon) [reply checks/branch_protection.go:60]: which comment?

File: checks/branch_protection.go
Comment (laurentsimon) [reply checks/branch_protection.go:403]: > Ah, I see. Then, let's calculate per-branch tier scores here and aggregate them later? So each branch will have:

{
  tier1: <score, passing>,
  tier2: <score, passing>,
 ...
}
which will be later aggregated across all branches. Would that work?

I think this would work. However, it goes against the idea of separating the check from score calculation because we now would be doing score calculation at the same time that we retrieve settings. The current code just retrieves data, and score calculation is done after, which is what we want in the long term. Btw, we also need the max score for each branch for sanity checking, I think.

File: checks/branch_protection.go
Comment (laurentsimon) [reply checks/branch_protection.go:39]: sg, will do. consider it done.

**Review by azeemshaikh38 — APPROVED** [review]

File: checks/branch_protection.go
Comment (azeemshaikh38) [reply checks/branch_protection.go:403]: Sg, lgtm

File: checks/branch_protection.go
Comment (asraa) [reply checks/branch_protection.go:60]: `Gated by requireStatusChecks=true` -- unless I'm misunderstanding!

File: docs/checks.md
Comment (asraa) [inline docs/checks.md:106]: suggest:
- Prevent force push
- Prevent branch deletion
- Include administrator for review

**Review by asraa — APPROVED** [review]
LGTM besides the small comment nits

**Comment by laurentsimon** [comment]
Thanks @asraa and @azeemsgoogle . Did not mean to be too pushy on this one... kinda came across this way though. @asraa please feel free to start an issue/PR to separate this check, if you're interested in tackling it.

File: docs/checks.md
Comment (laurentsimon) [reply docs/checks.md:106]: done, thanks.

File: checks/branch_protection.go
Comment (laurentsimon) [reply checks/branch_protection.go:60]: gotcha, will do. Thanks

**Comment by laurentsimon** [comment]
FYI ```Makefile:261: *** GITHUB_AUTH_TOKEN is undefined.  Stop.
Error: Process completed with exit code 2.```

**Comment by laurentsimon** [comment]
@naveensrinivasan @azeemshaikh38 please approve my integration tests. Thanks

**Comment by laurentsimon** [comment]
Note: the PR has 2 LGTMs but this does not trigger the automatic approval of the integration test :/

**Comment by laurentsimon** [comment]
Looks like I'm able to approve my own integration tests!

**Pull request merged**
